eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentRequest.scala

```diff
@@ -450,15 +450,17 @@ object PaymentRequest {
       case _ => 'm'
     }
 
-    def decode(input: String): Option[MilliSatoshi] =
-      input match {
-        case "" => None
-        case a if a.last == 'p' => Some(MilliSatoshi(a.dropRight(1).toLong / 10L)) // 1 pico-bitcoin == 10 milli-satoshis
-        case a if a.last == 'n' => Some(MilliSatoshi(a.dropRight(1).toLong * 100L))
-        case a if a.last == 'u' => Some(MilliSatoshi(a.dropRight(1).toLong * 100000L))
-        case a if a.last == 'm' => Some(MilliSatoshi(a.dropRight(1).toLong * 100000000L))
-        case a => Some(MilliSatoshi(a.toLong * 100000000000L))
+    def decode(input: String): Option[MilliSatoshi] = {
+      val amount = input match {
+        case "" => 0.msat
```

Comment:
I'm not sure I like the fact that you first convert to `0.msat` before converting to `None`.

Surely there is a more elegant way of doing this?

Reply:
I can instead change all the `Some(MilliSatoshi(a.toLong * multiplier))` to call an internal method that returns `None` if 0 msat and `Some` otherwise, would that be better?

Reply:
What about https://github.com/ACINQ/eclair/pull/1480/commits/19d3ae4790b2354a2a00faa6a07a07e8eda8a391 ?

Reply:
How about:

```scala
    def decode(input: String): Option[MilliSatoshi] = {
      val pattern = "([0-9]*)([pnum]?)".r
      input match {
        case pattern(a, _) if a.isEmpty || a.toLong == 0 => None
        case pattern(a, "p") => Some(MilliSatoshi(a.toLong / 10L)) // 1 pico-bitcoin == 10 milli-satoshis
        case pattern(a, "n") => Some(MilliSatoshi(a.toLong * 100L))
        case pattern(a, "u") => Some(MilliSatoshi(a.toLong * 100000L))
        case pattern(a, "m") => Some(MilliSatoshi(a.toLong * 100000000L))
        case pattern(a, "") => Some(MilliSatoshi(a.toLong * 100000000000L))
      }
    }
```

Reply:
Regex is evil, I know...

Reply:
I'd prefer avoiding a regex :)
What's wrong with the `flatMap`?

Reply:
Just aesthetics, but it's fine.